File: solutions/0261-graph-valid-tree/solution.java
import java.util.*;

  class Solution {
      public boolean validTree(int n, int[][] edges) {
          /*
              You have a graph of n nodes labeled from 0 to n - 1. You are given an integer n and a list of edges
  where edges[i] = [a_i, b_i] indicates that there is an undirected edge between nodes a_i and b_i in the graph

              Return true if the edges of the given graph make up a valid tree, and false otherwise

              Initial Approach

                  * Traverse through the tree and check for each adjacent node if another node is pointed to it

                      * I.E. in [[0,1],[1,2],[2,3],[1,3],[1,4]]
                          * Both 2 and 1 points to 3 etc

                      * If an adjacent implementation is detected, return false

          */

          if (edges.length != n - 1) {
              return false;
          }

          Map<Integer, List<Integer>> adjacencyList = new HashMap<>();

          for (int[] edge : edges) {
              adjacencyList.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
              adjacencyList.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
          }

          System.out.println(Arrays.deepToString(edges));

          for (Map.Entry<Integer, List<Integer>> entry : adjacencyList.entrySet()) {
              System.out.println(entry.getKey() + " -> " + entry.getValue());
          }

          Set<Integer> visited = new HashSet<>();

          // Check for cycles starting from node 0
          if (hasCycle(0, -1, adjacencyList, visited)) {
              return false;
          }

          // Check if all nodes are connected
          if (visited.size() != n) {
              return false;
          }

          return true;
      }

      private boolean hasCycle(int node, int parent, Map<Integer, List<Integer>> adjacencyList, Set<Integer>
  visited) {
          visited.add(node);
          List<Integer> neighbors = adjacencyList.getOrDefault(node, new ArrayList<>());

          for (int neighbor : neighbors) {
              if (!visited.contains(neighbor)) {
                  if (hasCycle(neighbor, node, adjacencyList, visited)) {
                      return true;
                  }
              } else if (neighbor != parent) {
                  return true;
              }
          }
          return false;
      }


    
    



  }

File: solutions/0261-graph-valid-tree/solution.py
class UnionFind:
    def __init__(self, elements):
        self.parents = {}
        self.rank = {}
        self.number_of_elements = elements

        for i in range(elements):
            # Set all parents to be their on parent
            self.parents[i] = i
            self.rank[i] = 0

    def find(self, element):
        while element != self.parents[element]:
            element = self.parents[element]
        return self.parents[element]

    def union(self, first_element, second_element):
        parent_one = self.find(first_element)
        parent_two = self.find(second_element)

        # Can't union nodes that have the same parent
        if parent_one == parent_two:
            return False
        
        if self.rank[parent_one] > self.rank[parent_two]:
            self.parents[parent_two] = parent_one
        else:
            self.parents[parent_one] = parent_two
            self.rank[parent_two] += 1
        self.number_of_elements -= 1
        return True
    
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
            
        union_find = UnionFind(n)

        for current_edge in edges:
            union_find.union(current_edge[0], current_edge[1])
        
        return union_find.number_of_elements == 1
    
